### PR TITLE
Introduce a Kafka Client DevUI component

### PR DESCRIPTION
**Pull request by alukin**
This PR adds the Kafka UI to the Quarkus development UI and allows you to add Kafka UI to your project as well.
By default, the UI is at the path `/q/kafka-ui`.
The UI is always available in the development mode if `quarkus-kafka-client` extension is used.

The Kafka UI allows you to

* view, create and delete topics
* view and create topic messages
* view consumer groups
* view access control lists
* view Kafka cluster info

### CODE REVIEW

File: extensions/kafka-client/ui/pom.xml
Comment (gsmet) [inline extensions/kafka-client/ui/pom.xml:53]:
```suggestion
                <!-- Move webapp to the proper location in resources -->
```

File: extensions/kafka-client/ui/README.asciidoc
Comment (gsmet) [inline extensions/kafka-client/ui/README.asciidoc:1]: I think we should drop that and rather include something in the documentation.

We don't have a similar thing for other extensions and my guess is that it will either be outdated or require duplicated work between the guide and this readme.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java
Comment (gsmet) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java:207]:
```suggestion
            // this mostly means that ACL controller is absent
```

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java
Comment (gsmet) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java:195]: Is it normal you have a TODO here? (it might be, just asking)

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java
Comment (gsmet) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java:31]: I don't see this one being used and I would rather drop it (the name doesn't follow our convention). If you need it then we need to discuss the name.

**Review by gsmet — DISMISSED** [review]
Thanks, that looks amazing.

I added a few very minor comments I noticed while scrolling through the diff.

Any chance you could put some screenshots in the description so people can have a quick look at it?

Thanks!

**Comment by gsmet** [comment]
I added the appropriate reviewers for deeper reviews/testing :).

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java:195]: yes
currently we do not have a generic response to transfer error messages to the UI, we just show a pop-up
we plan to fix it later on

File: extensions/kafka-client/ui/pom.xml
Comment (Nidum) [reply extensions/kafka-client/ui/pom.xml]: fixed

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaUiUtils.java]: fixed

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java
Comment (Nidum) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java]: fixed

File: extensions/kafka-client/ui/README.asciidoc
Comment (Nidum) [reply extensions/kafka-client/ui/README.asciidoc]: I reverted the commit with readme

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/model/response/KafkaMessagePage.java
Comment (ozangunalp) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/model/response/KafkaMessagePage.java:7]: BTW maybe it'd have been easier to handle the pagination if we hold the first and the last offsets of each partition contained in this message page.

**Review by ozangunalp — CHANGES_REQUESTED** [review]
Looks very good overall, I added some of my questions and suggestions in the comments.
The front-end part seems neat and component-based, considering we are not using any "modern" web frameworks.

It would have been great to add a smoke test for this.
And I still think that this whole UI can live inside a separate app and can be started with a dev service container.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (phillip-kruger) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:186]: Yes. This is what is used in the UIs and can be replaced by the user

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (phillip-kruger) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:176]: `config.js` is not used to customize the branding. It's a internal local file, but one that needs (maybe) some editing. It's like a poor man's templating, we replace certain `var`s in that file. Typically ones that can be configured.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (phillip-kruger) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:573]: No because this is not really Dev UI (It's a UI that is linked from Dev UI, like Swagger UI) You can not run a DevConsoleRoute outside of dev mode, where with this UI, you can enable that with a config (again like Swagger UI)

**Review by phillip-kruger — APPROVED** [review]
From a UI perspective - LGTM.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (alukin) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:566]: No, It is just mistake made during code cleanup. It will be removed, thank you.

**Comment by Nidum** [comment]
image
image
image
image
image
image
image
image
image

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaAdminClient.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaAdminClient.java:20]: The developers could use admin and topic clients separately, if they need. 
Admin client will be extended with RHOSAK features in the future, while TopicClient will remain as is, maybe with some improvements for messages.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaAdminClient.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaAdminClient.java:31]: fixed

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (Nidum) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:531]: fixed

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/AbstractHttpRequestHandler.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/AbstractHttpRequestHandler.java:55]: It's not an enum, I can't do that

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:36]: fixed

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:254]: let's postpone it till we have WS support

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/model/response/KafkaMessagePage.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/model/response/KafkaMessagePage.java:7]: renamed

to calculate a particular page we still will need to traverse through all the messages in every partition, as we can't know, how much messages we take for particular page from this particular pertition

but what we could do for improvement is cache some pages on the UI (like, first 3 for which the pagination is available right away)

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java
Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:227]: it's not guaranteed to be sorted

Comment (ozangunalp) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:227]: What I meant is when using the `getConsumerRecords` method the resulting list is sorted again.

Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:230]: because here we sort locally, within particular partition, and there - under multiple partitions, the merged result
here we have to sort to cut off what is redundant for partition, and there - for page

**Comment by phillip-kruger** [comment]
@alukin @Nidum make sure to build locally before commit, then the imports sound be sorted.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java
Comment (ozangunalp) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:230]: Ok, I didn't see that. I now realize that we are creating a consumer per topic-partition. 
Kafka consumer's unfairness between partitions really complicates things for this use case.
Do you think we can create a consumer per topic, possibly reusing it or only its client id?

Comment (Nidum) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:230]: no, we can't 
imagine if a user has two tabs open, or two users access this UI for a remote app. 
we enter same sode twice, shifting the offset to different position and getting wrong response in one of the pages.

Comment (ozangunalp) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java:230]: I understand that we can't reuse it between clients, I just thought creating a consumer per topic-partition was too much. Maybe there isn't a straightforward way of consuming multiple partitions in a fair way using a single consumer. We can revisit this later if needed.

**Review by ozangunalp — CHANGES_REQUESTED** [review]
I noticed two issues while testing the UI, maybe something got lost during the squash ?  
- The `bootstrap-multiselect.js` script tag is missing from the `index.html` file. The page fails to load without it. 
- `NodesPage` instance is not linked in the navigator, so the page won't get populated with fetched data.

Also sometimes I only see messages from the first partition of a multi-partition topic. I think it would be simpler to decouple the front-end pagination of messages from the actual message consumption.

**Comment by geoand** [comment]
The CI failures seem suspicious. Let me know if you need me to look into them

**Comment by Nidum** [comment]
> The CI failures seem suspicious. Let me know if you need me to look into them

@geoand, please, take a look, once you have time
Thank you!

**Comment by Nidum** [comment]
> I noticed two issues while testing the UI, maybe something got lost during the squash ?
> 
> * The `bootstrap-multiselect.js` script tag is missing from the `index.html` file. The page fails to load without it.
> * `NodesPage` instance is not linked in the navigator, so the page won't get populated with fetched data.
> 
> Also sometimes I only see messages from the first partition of a multi-partition topic. I think it would be simpler to decouple the front-end pagination of messages from the actual message consumption.

Fixed issues with multiselect and with missing link to node page, thank you!

Regarding messages - I am aware of this bug, I guess I can fix it pretty quickly

**Comment by geoand** [comment]
Once the domain experts approve the CI run, if there is still a failure, I'll have a look

**Comment by Nidum** [comment]
Cool, thank you!

**Comment by ozangunalp** [comment]
@geoand  Where those k8s tests fail is odd. I can't imagine how this PR can tamper with that. Let's give it a second go

**Comment by geoand** [comment]
Yeah, that's why I wanted to take a look

**Comment by Nidum** [comment]
@ozangunalp, should now show messages from all partitions on page load. 
Also, there was wrong count of messages shown - fixed this one as well.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (geoand) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:95]: Please try to avoid star imports

**Comment by geoand** [comment]
The reason the tests fail is because the `quarkus-kafka-client` now depends on `quarkus-vertx-http`. I am pretty sure that is not something we want

**Comment by phillip-kruger** [comment]
We can probably make that provided ? As when using the dev ui, it will be available...

**Comment by geoand** [comment]
We should not need to have to depend on the Vert.x HTTP extension for that. Just using the proper DevUI build items should be enough.

**Comment by phillip-kruger** [comment]
Yea, but the UI can also be used outside of Dev mode (like swagger UI)

**Comment by geoand** [comment]
Are you sure that's the intent? That's not what I gather reading the description?

**Comment by phillip-kruger** [comment]
Yes sure.

**Comment by geoand** [comment]
What's the rationale for having this in non-dev mode? The way I see it, it's not the same thing as a Swagger UI - this is something that affects the entire deployment and moreover is a read/write thing.

**Comment by ozangunalp** [comment]
@geoand What do you mean by "a read/write thing"? I agree that this is not really mature enough to be used in prod mode, let alone usefulness in prod mode is questionable. 
To reiterate, I really think that we should be packaging this as a separate app that we configure like a dev service to access the kafka broker configured for the application (dev service or otherwise).

**Comment by geoand** [comment]
> What do you mean by "a read/write thing"

I mean that with this we can not only view the state of system, but also alter it (`create and delete topics`, `create topic messages`).

> To reiterate, I really think that we should be packaging this as a separate app that we configure like a dev service to access the kafka broker configured for the application (dev service or otherwise).

My very strong preference here would be to start out with this being a DevUI component like other we have and if the need arises, think about moving this into a production-like feature.

**Comment by geoand** [comment]
I should not since I have not yet mentioned it, that I think this feature is amazing!
I just want us to be clear about its scope.

**Comment by ozangunalp** [comment]
There will be a need to evolve this to a standalone application, and the response to that was to include the component in prod-mode. To follow on: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/kafka.20work.20room/near/291290341

**Comment by ozangunalp** [comment]
> I should not since I have not yet mentioned it, that I think this feature is amazing! I just want us to be clear about its scope.

Yeah this is an awesome contribution by @alukin and @Nidum, we are just not sure how it should integrate into the ecosystem.

**Comment by geoand** [comment]
My personal opinion (as someone who has not been involved in this but who has plenty of context due to lengthly involvement in various aspects of Quarkus), is that it's better to start with a limited scope and expand it if necessary/requested, as opposed to starting really big and being unable to backtrack later on.

**Comment by ozangunalp** [comment]
@geoand so you are suggesting to keep this dev-mode only and integrate it through `DevConsoleRouteBuildItem` so that we don't have the quirks of pulling vertx-http as a dependency?

**Comment by geoand** [comment]
Yeah exactly, but the decision is ultimately up to you folks :).

Just keep in mind, that if we go down the current route (even using `provided`) we'll need to update the Kubernetes tests and consider this a breaking change because a kafka-client application will now be considered an HTTP application by our Kubernetes layer.

**Comment by ozangunalp** [comment]
+1

**Comment by phillip-kruger** [comment]
Maybe we move the UI component to not be automatically included? And the user needs to add it ?

**Comment by geoand** [comment]
That could be done, but it's really antithetical to the dev-mode philosophy where things just work with no configuration and then don't exist in production applications.

**Comment by phillip-kruger** [comment]
Then we do the Dev Services plan ? @ozangunalp ?

**Comment by ozangunalp** [comment]
@phillip-kruger can't we keep the code where it is now and replace the `quarkus-vertx-http` dependency by 
`quarkus-vertx-http-dev-console-spi`? I think we can expose the Json-RPC route via `DevConsoleRouteBuildItem`, but I don't know how we'd serve webjars.

**Comment by phillip-kruger** [comment]
I don't think that will work in non-dev mode though.

**Comment by cescoffier** [comment]
That looks amazing! Thank you so much!

**Comment by ozangunalp** [comment]
As @geoand mentioned kafka-client shouldn't pull vertx-http as dependency.
I'd like to merge this by making it dev-mode only, integrated fully into the current dev UI.
I'd like to propose minor changes to make to dev-console-spi in vertx-http extension so that an extension can expose its resources through a webjar description.

I didn't push onto the branch of this PR without discussing first, my change is here: 
https://github.com/ozangunalp/quarkus/commits/feature/kafka-client-dev-ui-squashed

**Comment by ozangunalp** [comment]
@phillip-kruger @cescoffier any thoughts on this?

**Comment by cescoffier** [comment]
@ozangunalp I think it's good to go.

**Comment by ozangunalp** [comment]
I agree, Let me rebase it again. One day I'll need to check again these continuous testing scenarios

**Review by ozangunalp — APPROVED** [review]

**Comment by ozangunalp** [comment]
@gsmet do you still have any objections for this one?

**Comment by gsmet** [comment]
Sorry, it just slipped to the second page and gets less love :). Let's get it in and get user feedback. Very nice contribution!

**Comment by ozangunalp** [comment]
Thanks @gsmet !

**Pull request merged**
